### PR TITLE
rgw/sfs: track db version, add metada upgrade path

### PR DESCRIPTION
**Pull request by jecluis**
This patch set is divided in two parts:
1. Keeping track of the database's version, via the `user_version` pragma; and,
2. Creating the bare bones for metadata upgrade between versions[^1].

It also removes the metadata schema tests, as that won't be needed anymore, given we can simply check against the version on startup.

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

[^1]: Which has been tested with the multipart v2 patch set, and works nicely.

### CODE REVIEW

**Comment by jecluis** [comment]
previous failures was an `EOF`:
```
[24/24] Linking CXX executable bin/unittest_rgw_sfs_sfs_bucket
/srv/ceph /srv/ceph
/srv/ceph
unexpected EOF
Error: Process completed with exit code 1.
```

File: src/rgw/driver/sfs/sqlite/dbconn.cc [hunk 1]
```diff
@@ -1,81 +1,83 @@
...
+      break;
     }
+
+    // example check to upgrade db
+    // if (cur_version == 1) {
+    //   upgrade_metadata_from_v1(db);
+    // }
+
+    storage.pragma.user_version(cur_version + 1);
   }
```
Comment (0xavi0) [inline src/rgw/driver/sfs/sqlite/dbconn.cc]: We can maybe do this without setting all the intermediate versions?

```c++
// ensure update path is possible
auto cur_version = get_version(ctt, storage);
ceph_assert(cur_version <= SFS_METADATA_VERSION);
ceph_assert(cur_version >= SFS_METADATA_MIN_VERSION);
// update
storage.pragma.user_version(SFS_METADATA_VERSION);
```

**Review by m-ildefons — DISMISSED** [review]
LGTM

**Review by 0xavi0 — COMMENTED** [review]
I think it could be good as a start point for upgrade management, but I would not drop the previous code for safety.

The problem I see is that if, for whatever reason (shit happens), we change the versions and db is not compatible... we'll drop data when calling `sync_schema`.

So I would complement this maybe with a test that keeps a previous version of the db and simply checks the result of `sync_schema` in a separate tool, or I would keep the code that copies the db and runs `sync_schema` on the copy first to ensure the result is not aggressive. (I think the later is safer)

The compatibility unit test was mainly checking that the result of `sync_schema` was the expected (`sqlite_orm` should be testing this already) and that the exception is caught. 
I agree is a pain to maintain that test, so we can maybe just drop the test for now, as it doesn't add much value and adds a lot of pain.

File: src/rgw/driver/sfs/sqlite/dbconn.cc (hunk 1 above)
Comment (jecluis) [reply src/rgw/driver/sfs/sqlite/dbconn.cc]: I just really like the idea of updating the db version as the upgrade goes on, in case something fails, so we know at what version we are at any given moment.

**Comment by jecluis** [comment]
> The problem I see is that if, for whatever reason (shit happens), we change the versions and db is not compatible... we'll drop data when calling `sync_schema`.

We can always run `sync_schema(true)`, i.e., with `preserve`. I think that way stuff won't be deleted.

**Comment by jecluis** [comment]
> previous failures was an `EOF`:
> 
> ```
> [24/24] Linking CXX executable bin/unittest_rgw_sfs_sfs_bucket
> /srv/ceph /srv/ceph
> /srv/ceph
> unexpected EOF
> Error: Process completed with exit code 1.
> ```

yet another failure with this. @m-ildefons any way we can debug this?

**Comment by 0xavi0** [comment]
> We can always run `sync_schema(true)`, i.e., with `preserve`. I think that way stuff won't be deleted.

That's another option, yeah, but I think this case would still be a problem:
From [sqliteorm docs](https://github.com/fnc12/sqlite_orm#migrations-functionality):

_if there are columns in storage that do not exist in db they will be added using 'ALTER TABLE ... ADD COLUMN ...' command and table data will not be dropped but if any of added columns is null but has not default value table will be dropped and recreated_

Same happens with the last option:
_if there is any column existing in both db and storage but differs by any of properties (type, pk, notnull) table will be dropped and recreated (dflt_value isn't checked cause there can be ambiguity in default values, please beware)._

 I would test those options before relying on `sync_schema(true)`, just in case. Because even in if copies the table to a temporary one, if it drops the original one, those objects won't be accessible anymore (I suspect db will also be corrupted due to FOREIGN key errors) 
In summary: I don't know if we can trust on `sync_schema` 100%

**Comment by jecluis** [comment]
Right, but this feels like something that will happen only if the schema is changed and the current version of the db is not updated. Because otherwise there should be a gate to prevent this from happening.

**Comment by 0xavi0** [comment]
> Right, but this feels like something that will happen only if the schema is changed and the current version of the db is not updated. Because otherwise there should be a gate to prevent this from happening.

Yup, it's just for that case in which we change the schema and forget to update the version and the reviewers don't realise. 
Unlikely "shit happens" scenario. 

It is also useful for detecting incompatible versions in the CI and find out if the release notes mention that metadata is no longer compatible. But we can move that to an external tool, I guess.

**Review by irq0 — DISMISSED** [review]
I like it! 

--- 

I think we can use this to build proper schema migration / database upgrade paths. I also think that sync_schema will be not enough sooner than later. A classic example would be splitting a field into two. We need to change schema, but also migrate data with custom queries. 

---

How about we store a timestamp in that integer (4 byte btw)? Could be useful for humans to tell how old a database version is and it is still monotonic.

**Comment by jecluis** [comment]
> I like it!
> 
> I think we can use this to build proper schema migration / database upgrade paths. I also think that sync_schema will be not enough sooner than later. A classic example would be splitting a field into two. We need to change schema, but also migrate data with custom queries.

Upgrade using sqlite directly is how I'm doing the upgrades -- couldn't find a good way to do it with `sqlite_orm` (see https://github.com/jecluis/ceph/commit/9504e5104ced58504e0f58e2f98582012f9ed5d8). And it works fine. :)


> How about we store a timestamp in that integer (4 byte btw)? Could be useful for humans to tell how old a database version is and it is still monotonic.

You mean like the first 4 bytes being schema version, last 4 bytes being a 4-byte timestamp?

**Review by 0xavi0 — APPROVED** [review]
lgtm
